### PR TITLE
build: add TeamCity build script to stress roachtests

### PR DESCRIPTION
**Pull request by erikgrinaker**
### roachtest: handle multiple runs of same test on TeamCity

The TeamCity markers output via `--teamcity` currently do not handle
multiple runs of the same test via `--count`, since they use the same
flow ID for all runs, making TeamCity unable to distinguish between
them. A similar problem exists with artifacts, where each run would
publish the root `artifacts/` directory rather than the `run_`
subdirectories.

This patch fixes the above issues by including run identifiers where
appropriate.

Release note: None

### roachtest: add roachtest/noop test

This adds a `roachtest/noop` test that instantly succeeds without
spinning up any servers, and a `roachtest/noop-maybefail` test that
randomly fails with 20% probability. These can be useful when testing
e.g. CI builds or other test infrastructure. They are disabled by
default and require the `roachtest` tag.

Release note: None

### roachtest: add --disable-issue flag to disable GitHub issues

Release note: None

### build: add TeamCity build script to stress roachtests

This adds a build script that can be used to stress roachtests on
TeamCity, in order to reproduce failures. The corresponding TeamCity
build is `Cockroach_Nightlies_RoachtestStress`.

Release note: None

### build: explicitly build binary for TeamCity nightly tests

The TeamCity nightly tests depend on the "publish bleeding edge" build,
which pulls in a ton of additional builds (e.g. for multiple OSes) that
we don't really need to run the nightlies. This can cause unrelated
build failures to fail the nightly tests.

This patch explicitly builds a CRDB binary in
`teamcity-nightly-roachtest.sh` such that the build can be decoupled
from "publish bleeding edge". This is a temporary mitigation until the
TeamCity builds and dependencies are cleaned up.

Touches https://cockroachlabs.atlassian.net/browse/DEVINF-16.
Touches cockroachdb#64263.

Release note: None

---

This is still a work-in-progress, but I suggest we merge the script itself
which makes it easier to generalize the TeamCity build and work on e.g.
branch specification.

There's a bit of work remaining on the actual build, which will be addressed
separately:

* Notifications: TeamCity does not expose the email address of the person
  who started the build, so for now we'll send notifications to the Slack
  channel `#roachtest-stress-ops`.

* Running clusters: We need to let the user know about any test clusters
  that are still running, so that these can be inspected and shut down as
  appropriate. We'll probably also need to change how roachtest extends
  cluster lifetimes, since we want to reuse clusters for tests but make sure
  every test invocation will extend the lifetime to 36 hours from the start
  of the test (not the cluster creation, which is currently the case).

* Trigger link: We should include a link on test failures to a TeamCity form
  for starting a stress build (if possible), or at least instructions for how
  to start one.

* Branch/commit selection: We need to make sure it's easy for users to pick
  which branch or commit to build.

### CODE REVIEW

File: build/teamcity-roachtest-stress.sh
Comment (tbg) [inline build/teamcity-roachtest-stress.sh:9]: it might be worth doing `export GCE_PROJECT=${GCE_PROJECT-andrei-jepsen}` here to avoid a stress run sucking up the quota needed for the nightlies.

File: build/teamcity-roachtest-stress.sh
Comment (tbg) [inline build/teamcity-roachtest-stress.sh:14]: It would be good to have this configurable (and default to false). It isn't always useful to leave the clusters around.

File: build/teamcity-roachtest-stress.sh
Comment (tbg) [inline build/teamcity-roachtest-stress.sh:21]: This does the right thing of using the auto-detection if BUILD_TAG is empty, right?

**Review by tbg — COMMENTED** [review]
Thanks for working on this! Before merging, I would suggest adding a dummy roachtest that does literally nothing and that's hidden behind a (roachtest) tag (i.e. doesn't have the default tag). We can invoke that with `--local` to work out most of the kinks you've described here.

> Test detection: TeamCity does not seem to handle the test output from
roachtest properly, since it will register one successful test even when
all of the tests fail. It also seems to believe the tests are nested.

That's odd, it does work for the nightly job. Or are you saying it doesn't work when `--count` is larger than one? That seems possible, we may just accidentally emit the closing teamcity markers outside of the loop rather than inside. In fact I bet that's it.

> Artifacts: There appears to be some sort of race where the artifacts are
removed before TeamCity is able to upload them. Artifacts are also
consistently missing for one of the test runs, but it seems to be random
which one is missing.

Hmm, I bet this is something similar. Look at what we tell TC to collect:

https://github.com/cockroachdb/cockroach/blob/fb55110da3e4963aa1ef869924f782753bf41c67/pkg/cmd/roachtest/z_test_runner.go#L462-L474

i.e. we tell TeamCity to collect `artifacts/TestFoo/*` but in that directory we have `run_1`, `run_2`, etc. So once the first test finishes we tell TC to collect that whole folder and I bet when we tell it to collect it again later it won't do so properly (perhaps thinking "I already did that!"). So either we tell it to collect everything only once, at the very end or, what is better, we make sure the artifactsSpec is minted for each run, i.e. the generated spec should be `artifacts/TestFoo/run_152/** => TestFoo/run_152` instead (and should be emitted after each run finishes, which makes sure that artifacts trickle in as they get finalized).

> Notifications: TeamCity does not expose the email address of the person
who started the build, so we'll have to send notifications from within the
build itself by querying the TeamCity REST API to get the email address.

I think it's fine to do nothing here for now. Or we set up Slack notifications for a `roachtest-stress-ops` channel for that job and someone who is stressing a roachtest and watching for failures can temporarily join that. One size fits all, gets the job done, little work, I like it. If people start using roachstress to a degree where this channel becomes too noisy, we are also certain to have other issues as well.

> Running clusters: We need to let the user know about any test clusters
that are still running, so that these can be inspected and shut down as
appropriate. We'll probably also need to change how roachtest extends
cluster lifetimes, since we want to reuse clusters for tests but make sure
every test invocation will extend the lifetime to 36 hours from the start
of the test (not the cluster creation, which is currently the case).

In another comment, I suggested leaving the default for `--debug` at false but making it configurable, I think that is still fine. Together with the #roachtest-stress-ops channel I don't think we need to be doing too much here, at least not until we see a problem.

> Nightly category: To get access to the nightly test secrets I had to place
the build in the nightly category, but it doesn't really belong there. We should
restructure how we manage secrets here.

This is likely best homed in dev-infs Jira at https://cockroachlabs.atlassian.net/secure/RapidBoard.jspa?rapidView=179&projectKey=DEVINF.

File: build/teamcity-roachtest-stress.sh
Comment (tbg) [inline build/teamcity-roachtest-stress.sh:13]: How is this preventing issues from getting posted for test failures?

File: build/teamcity-roachtest-stress.sh
Comment (tbg) [inline build/teamcity-roachtest-stress.sh:11]: How do you feel about making this happen for the main nightly jobs as well? We currently depend on publish bleeding edge but that's just a bad idea. I know it's a separate topic but we lose a roachtest run or two every week on that dependency, so while you're poking around here it might be a good idea to get rid of it.

There is something to be said about testing *exactly* the released binary, but I don't know, using the exact same build process ought to be enough for now given that dev-inf is not likely to disentangle the dependency for us anytime soon.

File: build/teamcity-roachtest-stress.sh
Comment (erikgrinaker) [reply build/teamcity-roachtest-stress.sh:21]: Yep.

File: build/teamcity-roachtest-stress.sh
Comment (erikgrinaker) [reply build/teamcity-roachtest-stress.sh:9]: Yeah, although maybe we should set up a properly named project for stuff like this instead of using `andrei-jepsen`?

File: build/teamcity-roachtest-stress.sh
Comment (erikgrinaker) [reply build/teamcity-roachtest-stress.sh]: True, done.

File: build/teamcity-roachtest-stress.sh
Comment (erikgrinaker) [reply build/teamcity-roachtest-stress.sh:11]: I suppose that should work, assuming there are no implicit dependencies on what that build does (which is a lot). I'd feel better about cleaning this up properly, or even just adding a minimal new canonical build that we could depend on here and expand for other builds, but as a temporary bandaid it should be fine.

File: build/teamcity-roachtest-stress.sh
Comment (erikgrinaker) [reply build/teamcity-roachtest-stress.sh:13]: I don't think it is, will look into it.

File: build/teamcity-roachtest-stress.sh
Comment (tbg) [reply build/teamcity-roachtest-stress.sh:11]: There are no implicit dependencies. If there are, I sure wouldn't mind knowing about them! I think we should move the main job to build its own but it doesn't have to be in this PR.

File: build/teamcity-roachtest-stress.sh
Comment (tbg) [reply build/teamcity-roachtest-stress.sh:9]: Agreed in principle, but that means filing a jira ticket against dev-inf and I wouldn't want to block this PR on that, so you need to put another project in as the default and `andrei-jepsen` is de facto used for that and only that.

File: build/teamcity-roachtest-stress.sh
Comment (erikgrinaker) [reply build/teamcity-roachtest-stress.sh:16]: Added a separate commit for this, we'll need to change the build once this is merged (possibly after a nightly run to make sure it works).

File: build/teamcity-roachtest-stress.sh
Comment (erikgrinaker) [reply build/teamcity-roachtest-stress.sh:9]: Makes sense, updated with `andrei-jepsen` and added https://cockroachlabs.atlassian.net/browse/DEVINF-140.

File: build/teamcity-roachtest-invoke.sh
Comment (tbg) [inline build/teamcity-roachtest-invoke.sh:2]: Do you want to keep the `-x` here? I suppose it doesn't hurt.

File: build/teamcity-roachtest-stress.sh
Comment (tbg) [reply build/teamcity-roachtest-stress.sh:17]: Want to give the line this comment is anchored on the same treatment as in the main script (redirect output, hide if successful)?

**Review by tbg — APPROVED** [review]

File: build/teamcity-roachtest-invoke.sh
Comment (erikgrinaker) [reply build/teamcity-roachtest-invoke.sh]: No strong opinion, mostly added it because the calling scripts used `-x` so I figured they'd want to see the final invocation. Remove it now.

File: build/teamcity-roachtest-stress.sh
Comment (erikgrinaker) [reply build/teamcity-roachtest-stress.sh]: Sure, done.

**Comment by erikgrinaker** [comment]
>>Test detection: TeamCity does not seem to handle the test output from
roachtest properly, since it will register one successful test even when
all of the tests fail. It also seems to believe the tests are nested.

>That's odd, it does work for the nightly job. Or are you saying it doesn't work when --count is larger than one? That seems possible, we may just accidentally emit the closing teamcity markers outside of the loop rather than inside. In fact I bet that's it.

It's because the flow identifiers used by the test runner does not distinguish between multiple runs of the same test:

https://github.com/cockroachdb/cockroach/blob/22c7022bbe05322279c09945c1bbb55ae23b7ed9/pkg/cmd/roachtest/test_runner.go#L620

The artifacts are a similar problem, as you pointed out. I'll add a commit here to make the test runner handle multiple test runs.

>> Notifications

>I think it's fine to do nothing here for now. Or we set up Slack notifications for a roachtest-stress-ops channel for that job and someone who is stressing a roachtest and watching for failures can temporarily join that.

Ok, I'll defer this for now.

>>Nightly category: To get access to the nightly test secrets I had to place
the build in the nightly category, but it doesn't really belong there. We should
restructure how we manage secrets here.

>This is likely best homed in dev-infs Jira

https://cockroachlabs.atlassian.net/browse/DEVINF-141

File: build/teamcity-roachtest-stress.sh
Comment (erikgrinaker) [reply build/teamcity-roachtest-stress.sh:22]: Added a commit with a `roachtest run --disable-issue` flag.

**Comment by erikgrinaker** [comment]
@tbg I think this should be ready for another look, I've addressed most comments/issues except the ones still listed in the PR description. Run in progress at https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestStress/3235092.

**Comment by tbg** [comment]
Nice!

image
image

**Comment by erikgrinaker** [comment]
Yeah, the annoying bit is that there's no easy way to map a failed test to the run ID (and thus artifacts/logs) for that test specifically. I added e.g. `#3` markers to the Go test output (i.e. `=== FAIL: roachtest/noop#3`), so it'll currently have to be dug out from the logs via those.

**Review by tbg — APPROVED** [review]
> Reviewed 3 of 4 files at r1, 1 of 1 files at r2, 9 of 9 files at r8, 4 of 4 files at r9, 3 of 3 files at r10, 2 of 4 files at r11, 2 of 2 files at r13, 1 of 1 files at r14.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67325)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker)

___
*[build/teamcity-roachtest-stress.sh, line 23 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/67325#-Mflhve74hA7K-JnoY6C:-Mflhve74hA7K-JnoY6D:b-bncwk2) ([raw file](https://github.com/cockroachdb/cockroach/blob/353d9a5520b168ffa1aa2bab984eaca6abe595e2/build/teamcity-roachtest-stress.sh#L23)):*
> ```Shell
> 
> build/builder/mkrelease.sh amd64-linux-gnu build bin/workload bin/roachtest bin/roachprod \
>   > artifacts/build.txt 2>&1 || cat artifacts/build.txt
> ```

this hides the error, right? You need something like `|| (cat artifacts/build.txt; exit 1)`?
___
*[pkg/cmd/roachtest/test_runner.go, line 477 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/67325#-MflewJz-6RWOgR-kcrO:-MflewJz-6RWOgR-kcrP:bhmy7xu) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bbfeaf704c49c398ff59db8ecee597d6fbebdf3/pkg/cmd/roachtest/test_runner.go#L477)):*
> ```Go
> 			if testToRun.runCount > 1 {
> 				artifactsSpec = fmt.Sprintf("%s/%s/** => %s/%s", base, runSuffix, escapedTestName, runSuffix)
> 			} else {
> ```

Why do we need this? Wouldn't things work if we made the first branch unconditional? I'd rather rely on a single rule to rule them all. (Or a comment explaining why that won't do).
___
*[pkg/cmd/roachtest/test_runner.go, line 626 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/67325#-MflhF2z1vTHb5B8pUMK:-MflhF2z1vTHb5B8pUML:b-vvpz8v) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bbfeaf704c49c398ff59db8ecee597d6fbebdf3/pkg/cmd/roachtest/test_runner.go#L626)):*
> ```Go
> 
> 	runID := t.Name()
> 	if runCount > 1 {
> ```

You could make this

```
if runNum > 1 {
  runID+= fmt.Sprintf("#%d", runNum)
}
```

and then we could stop passing `runCount` in here? Maybe it's fine, doesn't hurt to have it here.
___
*[pkg/cmd/roachtest/test_runner.go, line 684 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/67325#-Mflh0MbCNmFnsbYo3qS:-Mflh0MbCNmFnsbYo3qT:b-oigb3i) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bbfeaf704c49c398ff59db8ecee597d6fbebdf3/pkg/cmd/roachtest/test_runner.go#L684)):*
> ```Go
> 			r.maybePostGithubIssue(ctx, l, t, stdout, issueOutput)
> 		} else {
> 			shout(ctx, l, stdout, "--- PASS: %s (%s)", runID, durationStr)
> ```

This (and other messages of this kind) are trying to look like `go test`. I'm not saying we should ever try to parse it as `go test` output, but it might still be better to stick with `t.Name()` here.
___
*[pkg/cmd/roachtest/test_runner.go, line 898 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/67325#-MflheRrB58pmRFd-ZaZ:-MflheRrB58pmRFd-Za_:b42z6ap) ([raw file](https://github.com/cockroachdb/cockroach/blob/47a487f06428b1cd9feb47bde68b3a14a9cd91da/pkg/cmd/roachtest/test_runner.go#L898)):*
> ```Go
> func (r *testRunner) shouldPostGithubIssue(t test.Test) bool {
> 	opts := issues.DefaultOptionsFromEnv()
> 	return !r.config.disableIssue &&
> ```

We don't strictly need this, we could also just unset the github api token in the script. Just a suggestion in case you feel conflicted about adding a flag here.

**Review by tbg — APPROVED** [review]
Oh, that's why you did that. Maybe worth adding a noop-failure version of the test to check the UX?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67325#-:-MflikWn9OvRWnDJmE-p:b-a7v2ev)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker)

**Review by erikgrinaker — COMMENTED** [review]
>Oh, that's why you did that. Maybe worth adding a noop-failure version of the test to check the UX?

Yeah, but also because that's how `go test` outputs test status for subtests with duplicate or missing names.

Let me add another test that fails randomly.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67325#-:-Mfm3bQx4TSE2p16NAna:b-cx6lum)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[build/teamcity-roachtest-stress.sh, line 23 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/67325#-Mflhve74hA7K-JnoY6C:-Mfm3KXECFWyxPBWVn_9:bcvh3kh) ([raw file](https://github.com/cockroachdb/cockroach/blob/353d9a5520b168ffa1aa2bab984eaca6abe595e2/build/teamcity-roachtest-stress.sh#L23)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

this hides the error, right? You need something like `|| (cat artifacts/build.txt; exit 1)`?
</blockquote></details>

Good catch, just snagged this from some other build. Will fix.
___
*[pkg/cmd/roachtest/test_runner.go, line 477 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/67325#-MflewJz-6RWOgR-kcrO:-Mfm1Pm77f9Rul_ezcro:bwytt0n) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bbfeaf704c49c398ff59db8ecee597d6fbebdf3/pkg/cmd/roachtest/test_runner.go#L477)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Why do we need this? Wouldn't things work if we made the first branch unconditional? I'd rather rely on a single rule to rule them all. (Or a comment explaining why that won't do).
</blockquote></details>

Just to retain the current behavior for existing tests, in case we have tooling that depends on it. Happy to remove the branch if that's not the case.
___
*[pkg/cmd/roachtest/test_runner.go, line 626 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/67325#-MflhF2z1vTHb5B8pUMK:-Mfm1ZZ14TFHwpS8SM0z:b-ei4b3a) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bbfeaf704c49c398ff59db8ecee597d6fbebdf3/pkg/cmd/roachtest/test_runner.go#L626)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

You could make this

```
if runNum > 1 {
  runID+= fmt.Sprintf("#%d", runNum)
}
```

and then we could stop passing `runCount` in here? Maybe it's fine, doesn't hurt to have it here.
</blockquote></details>

Yeah, I just figured this would make it easier to identify run #1.
___
*[pkg/cmd/roachtest/test_runner.go, line 684 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/67325#-Mflh0MbCNmFnsbYo3qS:-Mfm1xLC8hTyJoqxti5c:bb0tmbm) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bbfeaf704c49c398ff59db8ecee597d6fbebdf3/pkg/cmd/roachtest/test_runner.go#L684)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

This (and other messages of this kind) are trying to look like `go test`. I'm not saying we should ever try to parse it as `go test` output, but it might still be better to stick with `t.Name()` here.
</blockquote></details>

Yeah, `runID` uses the same format as `go test` does when it runs subtests with unknown or duplicate names, e.g. `roachtest/noop#3`.

Currently, I believe these log lines are the only way to identify which specific run failed -- unless `roachtest` outputs some report at the very end. The TeamCity UI doesn't seem to have any way of distinguishing the failed run from the others.
___
*[pkg/cmd/roachtest/test_runner.go, line 898 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/67325#-MflheRrB58pmRFd-ZaZ:-Mfm2eqo8PUhnPZaSWqp:b-fp54ny) ([raw file](https://github.com/cockroachdb/cockroach/blob/47a487f06428b1cd9feb47bde68b3a14a9cd91da/pkg/cmd/roachtest/test_runner.go#L898)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

We don't strictly need this, we could also just unset the github api token in the script. Just a suggestion in case you feel conflicted about adding a flag here.
</blockquote></details>

True. I feel like explicit is better than implicit, but don't have a strong opinion here.

**Review by tbg — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67325)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker)

___
*[pkg/cmd/roachtest/test_runner.go, line 477 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/67325#-MflewJz-6RWOgR-kcrO:-Mfm66tg9sPaRX-HSyBl:b-z1gbha) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bbfeaf704c49c398ff59db8ecee597d6fbebdf3/pkg/cmd/roachtest/test_runner.go#L477)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Just to retain the current behavior for existing tests, in case we have tooling that depends on it. Happy to remove the branch if that's not the case.
</blockquote></details>

There shouldn't (the result should be indistinguishable) and if there is, I'd love to find out. Let's remove it.
___
*[pkg/cmd/roachtest/test_runner.go, line 684 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/67325#-Mflh0MbCNmFnsbYo3qS:-Mfm61mb1Ll5I2IoWxn6:b-di0ksa) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bbfeaf704c49c398ff59db8ecee597d6fbebdf3/pkg/cmd/roachtest/test_runner.go#L684)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Yeah, `runID` uses the same format as `go test` does when it runs subtests with unknown or duplicate names, e.g. `roachtest/noop#3`.

Currently, I believe these log lines are the only way to identify which specific run failed -- unless `roachtest` outputs some report at the very end. The TeamCity UI doesn't seem to have any way of distinguishing the failed run from the others.
</blockquote></details>

Ah, didn't realize this mimicks go test. Looks good then. (And independently).

**Review by tbg — APPROVED** [review]
Another thing - this doesn't have to be now but maybe you can add it since you're here - is to link to the script from the posted roachtest issues. Maybe something like this:

```diff
diff --git a/pkg/cmd/roachtest/test_runner.go b/pkg/cmd/roachtest/test_runner.go
index ada6d26793..0c08d8cb04 100644
--- a/pkg/cmd/roachtest/test_runner.go
+++ b/pkg/cmd/roachtest/test_runner.go
@@ -933,10 +933,16 @@ func (r *testRunner) maybePostGithubIssue(
 		Message:         msg,
 		Artifacts:       artifacts,
 		ExtraLabels:     labels,
-		ReproductionCommand: issues.ReproductionAsLink(
-			"roachtest README",
-			"https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/roachtest",
-		),
+		ReproductionCommand: func(r *issues.Renderer) {
+			issues.ReproductionAsLink(
+				"roachtest README",
+				"https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/roachtest",
+			)(r)
+			issues.ReproductionAsLink(
+				"CI job to stress roachtests",
+				"https://TODO",
+			)(r)
+		},
 	}
 	if err := issues.Post(
 		context.Background(),

```

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67325#-:-Mfm6LeA0NRnlU8zzcqy:b9mmcuk)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker)

**Comment by erikgrinaker** [comment]
>Another thing - this doesn't have to be now but maybe you can add it since you're here - is to link to the script from the posted roachtest issues. Maybe something like this:

I'm going to leave this until we've gotten the script merged, since I need to test that branch specification works, and for that I need to script to be available on `master` and some other branch.

**Comment by erikgrinaker** [comment]
Here's a test with some failures:

https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestStress/3240308?buildTab=tests&status=all

The UI seems a bit borked there, e.g. it reports 3 failed tests but shows 4 failed test runs (the correct failures are runs 8, 15, and 16). The TeamCity run numbers do not match the roachtest run numbers, presumably because TeamCity uses the order of service message outputs which are non-deterministic -- I suppose we could spawn tests in a deterministic order if we really wanted to have them match up.

Each failure includes failure messages from all test runs, I suppose we'd have to somehow tag every log line with the flow it belongs to (see [`out` service messages](https://www.jetbrains.com/help/teamcity/service-messages.html#Nested+test+reporting) and a related [`flowId` bug](https://youtrack.jetbrains.com/issue/TW-46515?_ga=2.113713223.1528990060.1627562696-2029485223.1617825768)), but I don't know if I want to go down that rabbit hole now.

Screenshot 2021-07-29 at 14 47 41

Seems like the message `test artifacts and logs in:` is the quickest way to figure out the failed test artifacts, the Go test output isn't used here. But in the raw build log they're still useful:

```
12:39:48   --- FAIL: roachtest/noop-maybefail#8 (0.00s)
12:39:48   test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/roachtest/noop-maybefail/run_8
12:39:48     roachtest.go:36,test_runner.go:777: randomly failing
```

**Comment by tbg** [comment]
Good enough for me! Feel free to merge this, for I'm about to sign off for the next two weeks. 🍹

**Comment by erikgrinaker** [comment]
>Good enough for me! Feel free to merge this, for I'm about to sign off for the next two weeks. 🍹

Yeah, I think this is workable for now. I'll let the nightly run with once with the new build script before removing the fat bleeding edge dep tomorrow, hopefully nothing breaks. 🤞

TFTR! 🍹

bors r=tbg

**Pull request merged**
